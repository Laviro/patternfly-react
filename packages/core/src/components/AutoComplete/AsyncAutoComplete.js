import React from 'react';
import { AsyncTypeahead } from 'react-bootstrap-typeahead';
import PropTypes from 'prop-types';

class AsyncAutoComplete extends React.Component {
  state = {
    options: this.props.options,
    isLoading: this.props.isLoading
  };

  onSearchStart = () => this.setState({ isLoading: true });

  onSearchEnd = options => this.setState({ options, isLoading: false });

  // query is the text string entered by the user.
  handleSearch = query => {
    this.onSearchStart();
    Promise.resolve(this.props.onSearch(query)).then(options =>
      this.onSearchEnd(options)
    );
  };

  render = () => (
    <AsyncTypeahead
      {...this.props}
      onSearch={this.handleSearch}
      options={this.state.options}
      isLoading={this.state.isLoading}
    />
  );
}

AsyncAutoComplete.propTypes = {
  onSearch: PropTypes.func.isRequired,
  options: PropTypes.array,
  isLoading: PropTypes.bool,
  delay: PropTypes.number,
  promptText: PropTypes.object,
  searchText: PropTypes.object,
  useCache: PropTypes.bool
};

AsyncAutoComplete.defaultProps = {
  options: [],
  isLoading: false,
  delay: 200,
  promptText: <span>Type to search...</span>,
  searchText: <span>Searching...</span>,
  useCache: true
};

export default AsyncAutoComplete;
