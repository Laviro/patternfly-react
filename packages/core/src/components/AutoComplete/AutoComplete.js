import React from 'react';
import { Typeahead } from 'react-bootstrap-typeahead';
import PropTypes from 'prop-types';

const AutoComplete = props => <Typeahead {...props} />;
AutoComplete.propTypes = {
  align: PropTypes.string,
  allowNew: PropTypes.bool,
  bsSize: PropTypes.string,
  caseSensitive: PropTypes.bool,
  clearButton: PropTypes.bool,
  defaultSelected: PropTypes.arrayOf(PropTypes.string),
  disabled: PropTypes.bool,
  dropup: PropTypes.bool,
  highlightOnlyResult: PropTypes.bool,
  filterBy: PropTypes.oneOfType([PropTypes.func, PropTypes.array]),
  labelKey: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  maxHeight: PropTypes.string,
  maxResults: PropTypes.number,
  minLength: PropTypes.number,
  multiple: PropTypes.bool,
  options: PropTypes.array.isRequired,
  placeholder: PropTypes.string,
  selectHintOnEnter: PropTypes.bool
};

AutoComplete.defaultProps = {
  align: 'justify',
  allowNew: false,
  bsSize: null,
  caseSensitive: false,
  defaultSelected: [],
  clearButton: false,
  disabled: false,
  dropup: false,
  filterBy: [],
  highlightOnlyResult: false,
  labelKey: 'label',
  maxHeight: '300px',
  maxResults: 100,
  minLength: 0,
  multiple: false,
  placeholder: null,
  selectHintOnEnter: false
};
export default AutoComplete;
