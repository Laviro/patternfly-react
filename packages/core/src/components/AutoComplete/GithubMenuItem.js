import React from 'react';
import PropTypes from 'prop-types';

const GithubMenuItem = props => (
  <div key={props.option.id}>
    <img
      alt="avatar"
      height="20px"
      src={props.option.avatar_url}
      style={{ borderRadius: '10px', margin: '5px' }}
    />
    {props.option.login}
  </div>
);

GithubMenuItem.propTypes = {
  option: PropTypes.object.isRequired
};

export default GithubMenuItem;
