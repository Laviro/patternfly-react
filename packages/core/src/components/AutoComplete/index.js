export AutoComplete from './AutoComplete';
export AsyncAutoComplete from './AsyncAutoComplete';
