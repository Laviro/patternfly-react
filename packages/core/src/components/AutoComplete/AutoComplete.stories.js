import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, select, boolean, number } from '@storybook/addon-knobs';
import {
  storybookPackageName,
  STORYBOOK_CATEGORY
} from 'storybook/constants/siteConstants';
import { name } from '../../../package.json';
import AutoComplete from './AutoComplete';
import AsyncAutoComplete from './AsyncAutoComplete';
import { Form, FormGroup, Button } from '../../index';
import Mocks from './mock';
import GithubMenuItem from './GithubMenuItem';

const AutoCompleteStories = storiesOf(
  `${storybookPackageName(name)}/${STORYBOOK_CATEGORY.WIDGETS}/AutoComplete`,
  module
);

AutoCompleteStories.addDecorator(withKnobs);

AutoCompleteStories.addWithInfo('Menu Alignment', () => {
  const align = select('Alignment', [undefined, 'justify', 'left', 'right']);

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Menu Alignment</h2>
      <AutoComplete
        align={align}
        options={Mocks.countries}
        labelKey="country"
        placeholder="Choose a state..."
      />
    </div>
  );
});

AutoCompleteStories.addWithInfo('Input Size', () => {
  const bsSize = select('Size', [undefined, 'small', 'default', 'large']);

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Input Size</h2>
      <AutoComplete
        bsSize={bsSize}
        options={Mocks.countries}
        labelKey="name"
        placeholder="Choose a state..."
      />
    </div>
  );
});

AutoCompleteStories.addWithInfo('Multiple Selection', () => {
  const clearButton = boolean('Clear Button', true);
  const multiple = boolean('Multiple Selection', true);

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Input Size</h2>
      <AutoComplete
        multiple={multiple}
        clearButton={clearButton}
        selected={['Burkina Fasu', 'Papua New Guinea', 'Satellite']}
        options={Mocks.countries}
        labelKey="name"
        placeholder="Choose a state..."
      />
    </div>
  );
});

AutoCompleteStories.addWithInfo('Highlight Results', () => {
  const highlightOnlyResult = boolean('highlight Only Result', true);
  const selectHintOnEnter = boolean('select Hint On Enter', true);

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Highlight Results</h2>
      <AutoComplete
        highlightOnlyResult={highlightOnlyResult}
        selectHintOnEnter={selectHintOnEnter}
        options={Mocks.countries}
        labelKey="name"
        placeholder="Choose a state..."
      />
    </div>
  );
});

AutoCompleteStories.addWithInfo('Allow Adding New Items', () => {
  const allowNew = boolean('Allow Adding New Items', true);

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Allow Adding New Items</h2>
      <p>
        <strong>
          If a search result doesn&apos;t exist, allow the user to add a custom
          input.
        </strong>
      </p>
      <AutoComplete
        multiple
        allowNew={allowNew}
        options={Mocks.countries}
        labelKey="name"
        placeholder="Choose a state..."
      />
    </div>
  );
});

AutoCompleteStories.addWithInfo('Min Chars To AutoComplete', () => {
  const minLength = number('Set the Minimum Length to Autocomplete', 1);

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>AutoCompletion will start when there are enough characters.</h2>
      <AutoComplete
        minLength={minLength}
        options={Mocks.countries}
        labelKey="name"
        placeholder="Choose a state..."
      />
    </div>
  );
});

AutoCompleteStories.addWithInfo('Use With Forms', () => {
  const submitFormOnEnter = boolean('Submit Form On Enter', true);
  const handleSubmit = e => {
    alert('Form was submitted successfully !');
    e.preventDefault();
  };
  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Submit Form On &quot;Enter&quot;</h2>
      <p>
        <strong>Press the &quot;Enter&quot; key</strong>
      </p>
      <Form horizontal onSubmit={handleSubmit}>
        <div className="col-sm-10">
          <FormGroup>
            <AutoComplete
              minLength={1}
              submitFormOnEnter={submitFormOnEnter}
              labelKey="name"
              options={Mocks.countries}
              placeholder="Choose a state..."
            />
          </FormGroup>
        </div>
        <div className="col-sm-2">
          <FormGroup>
            <Button bsStyle="primary" type="submit">
              Submit
            </Button>
          </FormGroup>
        </div>
      </Form>
    </div>
  );
});

AutoCompleteStories.addWithInfo('Using Async Calls', () => {
  // query is the text string entered by the user.
  const handleSearch = query =>
    fetch('https://api.github.com/repos/patternfly/patternfly-react/forks')
      .then(response => response.json())
      .then(forks => forks.map(fork => fork.owner));

  return (
    <div className="container">
      <br />
      <br />
      <br />
      <h2>Search a Forker</h2>
      <AsyncAutoComplete
        labelKey="login"
        minLength={0}
        placeholder="Search someone who have forked Patternfly-react.."
        onSearch={handleSearch}
        multiple
        clearButton
        highlightOnlyResult
        renderMenuItemChildren={option => <GithubMenuItem option={option} />}
      />
    </div>
  );
});
