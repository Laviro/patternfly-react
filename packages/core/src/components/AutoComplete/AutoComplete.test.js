import React from 'react';
import { shallow } from 'enzyme';
import AutoComplete from './AutoComplete';

test('AutoComplete is working !!', () => {
  const component = shallow(
    <AutoComplete
      clearButton
      multiple
      allowNew
      options={['One', 'Two', 'Three']}
    />
  );

  expect(component).toMatchSnapshot();
});
