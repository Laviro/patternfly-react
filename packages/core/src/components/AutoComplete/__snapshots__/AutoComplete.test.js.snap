// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete is working !! 1`] = `
<OnClickOutside(TypeaheadContainer(WrappedTypeahead))
  align="justify"
  allowNew={true}
  bsSize={null}
  caseSensitive={false}
  clearButton={true}
  defaultSelected={Array []}
  disabled={false}
  dropup={false}
  eventTypes={
    Array [
      "mousedown",
      "touchstart",
    ]
  }
  excludeScrollbar={false}
  filterBy={Array []}
  highlightOnlyResult={false}
  labelKey="label"
  maxHeight="300px"
  maxResults={100}
  minLength={0}
  multiple={true}
  options={
    Array [
      "One",
      "Two",
      "Three",
    ]
  }
  outsideClickIgnoreClass="ignore-react-onclickoutside"
  placeholder={null}
  preventDefault={false}
  selectHintOnEnter={false}
  stopPropagation={false}
/>
`;
